examples/aligned_accessor/aligned_accessor.cpp

```diff
@@ -0,0 +1,714 @@
+// Otherwise, use a compiler-specific equivalent if available.
+
+// NOTE (mfh 2022/08/08) BYTE_ALIGNMENT must be unsigned and a power of 2.
+#if defined(__cpp_lib_assume_aligned)
+#  define _MDSPAN_ASSUME_ALIGNED( ELEMENT_TYPE, POINTER, BYTE_ALIGNMENT ) (std::assume_aligned< BYTE_ALIGNMENT >( POINTER ))
+  constexpr char assume_aligned_method[] = "std::assume_aligned";
+#elif defined(__ICL)
+#  define _MDSPAN_ASSUME_ALIGNED( ELEMENT_TYPE, POINTER, BYTE_ALIGNMENT ) POINTER
+  constexpr char assume_aligned_method[] = "(none)";
+#elif defined(__ICC)
+#  define _MDSPAN_ASSUME_ALIGNED( ELEMENT_TYPE, POINTER, BYTE_ALIGNMENT ) POINTER
+  constexpr char assume_aligned_method[] = "(none)";
+#elif defined(__clang__)
+#  define _MDSPAN_ASSUME_ALIGNED( ELEMENT_TYPE, POINTER, BYTE_ALIGNMENT ) POINTER
+  constexpr char assume_aligned_method[] = "(none)";
+#elif defined(__GNUC__)
+  // __builtin_assume_aligned returns void*
+#  define _MDSPAN_ASSUME_ALIGNED( ELEMENT_TYPE, POINTER, BYTE_ALIGNMENT ) reinterpret_cast< ELEMENT_TYPE* >(__builtin_assume_aligned( POINTER, BYTE_ALIGNMENT ))
+  constexpr char assume_aligned_method[] = "__builtin_assume_aligned";
+#else
+#  define _MDSPAN_ASSUME_ALIGNED( ELEMENT_TYPE, POINTER, BYTE_ALIGNMENT ) POINTER
+  constexpr char assume_aligned_method[] = "(none)";
+#endif
+
+// Some compilers other than Clang or GCC like to define __clang__ or __GNUC__.
+// Thus, we order the tests from most to least specific.
+#if defined(__ICL)
+#  define _MDSPAN_ALIGN_VALUE_ATTRIBUTE( BYTE_ALIGNMENT ) __declspec(align_value( BYTE_ALIGNMENT ));
+  constexpr char align_attribute_method[] = "__declspec(align_value(BYTE_ALIGNMENT))";
+#elif defined(__ICC)
+#  define _MDSPAN_ALIGN_VALUE_ATTRIBUTE( BYTE_ALIGNMENT ) __attribute__((align_value( BYTE_ALIGNMENT )));
+  constexpr char align_attribute_method[] = "__attribute__((align_value(BYTE_ALIGNMENT)))";
+#elif defined(__clang__)
+#  define _MDSPAN_ALIGN_VALUE_ATTRIBUTE( BYTE_ALIGNMENT ) __attribute__((align_value( BYTE_ALIGNMENT )));
+  constexpr char align_attribute_method[] = "__attribute__((align_value(BYTE_ALIGNMENT)))";
+#else
+#  define _MDSPAN_ALIGN_VALUE_ATTRIBUTE( BYTE_ALIGNMENT )
+  constexpr char align_attribute_method[] = "(none)";
+#endif
+
+template<class ElementType>
+constexpr bool
+valid_byte_alignment(const std::size_t byte_alignment,
+		     const std::size_t size_of_element)
+{
+  return byte_alignment >= alignof(ElementType); // && is a power of two
+}
+
+template<class ElementType, std::size_t byte_alignment>
+struct aligned_pointer {
+  // x86-64 ICC 2021.5.0 emits warning #3186 ("expected typedef declaration") here.
+  // No other compiler (including Clang, which has a similar type attribute) has this issue.
+  using type = ElementType* _MDSPAN_ALIGN_VALUE_ATTRIBUTE( byte_alignment );
+};
+
+template<class ElementType, std::size_t byte_alignment>
+using aligned_pointer_t = typename aligned_pointer<ElementType, byte_alignment>::type;
```

Comment:
Why do you bother with the trait and don't directly define the alias?

Reply:
Good point -- I must have mixed up in my head some missing C++14 thing from this C++11 feature.  I'll change it.

Reply:
Fixed; thanks!

Reply:
I ended up bringing back the struct, because this lets me check alignment at compile time for all uses of the alias.  (Just defining the alias directly wouldn't let me impose a `static_assert` on it for all specializations.)